Save Grid state in MS ProtectedBrowserStorage (#126)
telerik/blazor-ui: grid/save-state-in-ProtectedBrowserStorage/readme.md

# Store Grid State in ASP.NET Core Protected Browser Storage

> The <a href="https://docs.telerik.com/blazor-ui/components/grid/state" target="_blank">Telerik Grid State feature</a> lets you get and set the grid state with code. It also lest you store that state for your users so they can get back where they left off (filters, page index, sorting, and so on).

Reviewer: ```suggestion
The <a href="https://docs.telerik.com/blazor-ui/components/grid/state" target="_blank">Telerik Grid State feature</a> lets you get and set the grid state with code. It also lets you store that state for your users so they can get back where they left off (filters, page index, sorting, and so on).
```

> The grid provides you with a `GridStat<T>` object with the state that contains that information so the application can choose where to store it. Common choices are a database or the browser local storage. Both of these options require a string value, so the object from the grid must be serialized. The `System.Text.Json` serializer that comes with the framework is supported.

Reviewer: ```suggestion
The Grid provides you with a `GridState<T>` object with the state information, so the application can choose where to store it. Common choices are a database or the browser's local storage. Both of these options require a string value, so the object from the grid must be serialized. The `System.Text.Json` serializer that comes with the framework is supported.
```

> This sample project shows how to use the `Microsoft.AspNetCore.Components.Server.ProtectedBrowserStorage` to store the grid state in the browser in an encrypted format (the framework does the encryption). You can read more on using this framework service in the following documentation page: <a href="https://docs.microsoft.com/en-us/aspnet/core/blazor/state-management" target="_blank">https://docs.microsoft.com/en-us/aspnet/core/blazor/state-management</a>.

Reviewer: ```suggestion
This sample project shows how to use the `Microsoft.AspNetCore.Components.Server.ProtectedBrowserStorage` to store the Grid state in the browser in an encrypted format (the framework does the encryption). You can read more about this framework service in the following documentation page: <a href="https://docs.microsoft.com/en-us/aspnet/core/blazor/state-management" target="_blank">https://docs.microsoft.com/en-us/aspnet/core/blazor/state-management</a>.
```

> In this sample project, there are two pages:

Reviewer: ```suggestion
There are two pages in the project:
```

- `Index.razor` uses the framework service directly and lets it serialize the data.

> - `ExplicitSerialization.razor` serializes the data explicitly to a string before passing it to the framework service. This can be useful for debugging purposes in case you face issues or the built-in serialization approach in the service changes/breaks.

Reviewer: ```suggestion
- `ExplicitSerialization.razor` serializes the data explicitly to a string before passing it to the framework service. This can be useful for debugging purposes in case you face issues, or the built-in serialization approach in the service changes/breaks.
```

## Notes

At the time of writing, there are two caveats with this service:

> * It seems like only injecting the framework service into a component is possible and if you need to make it reusable, the approach from the official documentation is to use a cascading parameter from a parent component. My attempt to inject it into another service was not successful.

Reviewer: ```suggestion
* The framework service needs to be injected into every Razor component that uses it. The official Microsoft documentation suggests reusing the service via a cascading parameter from a parent component. My attempt to inject it into another service was not successful.
```
